src/message_bus.zig

```diff
@@ -587,245 +621,273 @@ const Connection = struct {
 
-        const bytes_received = result catch |err| {
-            // TODO: maybe don't need to close on *every* error
-            log.err("error receiving body from {}: {}", .{ self.peer, err });
-            self.shutdown();
-            return;
-        };
+        if (header.size < @sizeOf(Header) or header.size > config.message_size_max) {
+            log.err("invalid header size {d} received from {}", .{ header.size, self.peer });
+            self.terminate(.shutdown);
+            return null;
+        }
 
-        // No bytes received means that the peer closed the connection.
-        if (bytes_received == 0) {
-            // Skip the shutdown syscall and go straight to close.
-            self.state = .shutting_down;
-            self.maybe_close();
-            return;
+        if (header.cluster != self.message_bus.server.cluster) {
+            log.err("message addressed to the wrong cluster: {}", .{header.cluster});
+            self.terminate(.shutdown);
+            return null;
         }
 
-        self.recv_progress += bytes_received;
+        self.set_peer(header);
 
-        if (self.recv_progress < @sizeOf(Header)) {
-            // The header has not yet been fully received.
-            self.recv_header();
-            return;
+        if (data.len < header.size) {
+            self.get_recv_message_and_recv();
+            return null;
         }
-        assert(self.recv_progress == @sizeOf(Header));
 
-        const header_received = self.incoming_message.?.header;
-        if (!header_received.valid_checksum()) {
-            log.err("invalid checksum on header received from {}", .{self.peer});
-            self.shutdown();
-            return;
+        const body = data[@sizeOf(Header)..header.size];
+        if (!header.valid_checksum_body(body)) {
+            log.err("invalid body checksum received from {}", .{self.peer});
+            self.terminate(.shutdown);
+            return null;
         }
 
+        self.recv_parsed += header.size;
+
+        // Return the parsed message using zero-copy if we can, or copy if the client is pipelining:
+        // If this is the first message but there are messages in the pipeline then we copy the
+        // message so that its sector padding (if any) will not overwrite the front of the pipeline.
+        // If this is not the first message then we must copy the message to a new message as each
+        // message needs to have its own unique `references` and `header` metadata.
+        if (self.recv_progress == header.size) return self.recv_message.?.ref();
+
+        const message = self.message_bus.get_message() orelse {
+            log.notice("waiting for a free message for pipelined message from {}", .{self.peer});
+            // TODO Can this next line cause an assertion failure in `set_recv_message_and_recv()`?
+            self.recv_parsed -= header.size;
+            self.get_recv_message_and_recv();
+            return null;
```

Comment:
I'm pretty sure this will cause a busy loop if there are no messages available and the recv buffer is full. I think we should just drop the message if there is no `Message` available.

Reply:
Yes, this was also the thing that bothered me.

Reply:
I like the `parse_messages()` and `parse_message()` split and think we should try hard to keep it, solving the issue now where our `?*Message` return from `parse_message()` is not enough to signal that we want to drop the message and keep parsing more. I'm also thinking if there are no messages available, it's probably also a good idea to terminate the connection, especially if it's a client connection.

Reply:
> I'm also thinking if there are no messages available, it's probably also a good idea to terminate the connection, especially if it's a client connection.

Hmm, I hadn't thought of that yet. That would significantly simplify this code and make the fast path less branchy. I think if we do this we should treat client and replica connections the same though, otherwise we don't really benefit in terms of code complexity.

Reply:
> Regardless, the TODO comment here can be removed I think.

I'm not sure. What if `recv_parsed` is 0, and we have more messages in the pipeline, so that we have to copy the first message, but there are no free messages available?

Also, if there are no free messages available, we may also have `recv_progress == message_size_max` and then try to `recv()` more which would be impossible (we now have an assertion for that).

Options for solving this so far:

1. Drop message, but that breaks our parse_message() interface.
2. Terminate the connection, dropping all messages for it. Probably what we want to do especially for clients.
3. get_recv_message_and_recv(), system does more copies when it's already busy, assertion in `set_recv_message_and_recv()` must be loosened, and we shouldn't always issue impossible `recv()`.
4. Find another way here to wait for a message, without the issues of option 2.

Reply:
I don't think inlining `parse_message()` into `parse_messages()` is that big a deal, `parse_messages()` quite short currently. If we don't want to drop replica connections on running out of messages, that would be my preferred solution.

If we allocate enough messages up front, we should never end up dropping replica or client connections. I think that this would be the solution that leads to the overall simplest code. However, it migth leave us vulnerable to reconnect loops and inefficiency when operating on the edge of the message limit. I think it is possible for us to determine a true upper limit on how many messages the entire system can possibly use at any one time though, which would eliminate this issue.

Reply:
> If we allocate enough messages up front, we should never end up dropping replica or client connections. I think that this would be the solution that leads to the overall simplest code. However, it migth leave us vulnerable to reconnect loops and inefficiency when operating on the edge of the message limit. I think it is possible for us to determine a true upper limit on how many messages the entire system can possibly use at any one time though, which would eliminate this issue.

100% agreed on this.

Let's just terminate the connection, dropping all messages, for client or replica connections, for simplicity and also to reduce load on the system.

And then at the same time, we also decrease the probability of catastrophe by:
1. leaning more heavily into static allocation and determining a true upper limit, and
2. optimizing `MessageBus.get_message()` to accept a `size_max` parameter, so that it can return a header-only message when that's all we need, to improve the chances that we don't run out of messages.

Reply:
Sounds like a plan to me. I'll get started on the implementation unless you're itching to write it :)

Reply:
Great, you go for it! I'll just push the commit to terminate here with a TODO for those two strategies.